File: eval/eval_kitti.m
% Copyright 2017. All rights reserved.
% Computer Vision Group, Visual Computing Institute
% RWTH Aachen University, Germany
% This file is part of the semodepth project.
% Authors: Yevhen Kuznietsov (yevkuzn@gmail.com OR Yevhen.Kuznietsov@esat.kuleuven.be)
% semodepth project is free software; you can redistribute it and/or modify it under the
% terms of the GNU General Public License as published by the Free Software
% Foundation; either version 3 of the License, or any later version.
% semodepth project is distributed in the hope that it will be useful, but WITHOUT ANY
% WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
% PARTICULAR PURPOSE. See the GNU General Public License for more details.


%Set paths to gt and to predicted depth maps
gt_path = '...';
pred_path = '...';

%Get list of all files in gt / prediction directories
gt_files = dir(fullfile(gt_path, '*.mat'));
pred_files = dir(fullfile(pred_path, '*.mat'));

%Initialize cumulative metric values
crms = 0;
n_pxls = 0; %Number of valid pixels
crmsl = 0;
cabs_rel = 0;
csq_rel = 0;

cacc1 = 0;
cacc2 = 0;
cacc3 = 0;

%Evaluate
for i = 1:1:numel(gt_files)
    %Load GT and corresponding predictions
    gt = load(fullfile(gt_path, gt_files(i).name));
    pred = load(fullfile(pred_path, gt_files(i).name));
    gt_depths = double(gt.depth);
    [h, w] = size(gt_depths);
    pred_depths = reshape(pred.mat(:), 96, 320);
    %Resize predictions to match GT resolution
    predictions = double(imresize(pred_depths, [h, w], 'bilinear'));
    %Cap predicted values (use either 80 or 50 m)
    predictions(predictions > 80.0) = 80.0;
    predictions(predictions < 1.0) = 1.0;
    %Shows number of processed images
    if mod(i, 100) == 0
        disp(i);
    end;
  
    mask = gt_depths > 0.01; %All pixels with valid GT
    crop = zeros(h, w);
    %Garg / Eigen crop for image resolution 370x1224
    crop(fix(h-218):fix(h-3), 44 :1180) = 1;
    mask = mask & logical(crop); %All pixels to be used for evaluation
    n_pxls = n_pxls + sum(mask(:)); 
    
    %Compute RMSE
    rms = (gt_depths(:) - predictions(:)).^2;
    rms(~mask) = 0;
    crms = crms + sum(rms);
    
    %Compute RMSE log
    rmsl = (log(gt_depths(:)) - log(predictions(:))).^2;
    rmsl(~mask) = 0;
    crmsl = crmsl + sum(rmsl);
    
    gt_depths(gt_depths < 0.5) = 0.5; %epsilon
    
    %Compute ARD
    abs_rel = abs(gt_depths(:) - predictions(:)) ./ gt_depths(:);
    abs_rel(~mask) = 0;
    cabs_rel = cabs_rel + sum(abs_rel);
    
    %Compute SRD
    sq_rel = ((gt_depths(:) - predictions(:)).^2) ./ gt_depths(:);
    sq_rel(~mask) = 0;
    csq_rel = csq_rel + sum(sq_rel);
    
    max_ratio = max(gt_depths ./ predictions, predictions ./ gt_depths);
    
    %Compute accuracies
    acc1 = double(max_ratio < 1.25 & mask);
    acc2 = double(max_ratio < 1.25^2 & mask);
    acc3 = double(max_ratio < 1.25^3 & mask);
    
    cacc1 = cacc1 + sum(acc1(:));
    cacc2 = cacc2 + sum(acc2(:));
    cacc3 = cacc3 + sum(acc3(:));
end


RMSE = sqrt(crms / n_pxls)
RMSE_log = sqrt(crmsl / n_pxls)
ABS_REL = cabs_rel / n_pxls
SQ_REL = csq_rel / n_pxls
ACCURACY1 = cacc1 / n_pxls
ACCURACY2 = cacc2 / n_pxls
ACCURACY3 = cacc3 / n_pxls

File: eval/eval_kitti.py
# Copyright 2017. All rights reserved.
# Computer Vision Group, Visual Computing Institute
# RWTH Aachen University, Germany
# This file is part of the semodepth project.
# Authors: Yevhen Kuznietsov (yevkuzn@gmail.com OR Yevhen.Kuznietsov@esat.kuleuven.be)
# semodepth project is free software; you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation; either version 3 of the License, or any later version.
# semodepth project is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
# PARTICULAR PURPOSE. See the GNU General Public License for more details.


import os
import locale
import scipy.io as sio
import numpy as np
import cv2

#For sorting files in alphabetical order
locale.setlocale(locale.LC_ALL, 'en_US.UTF-8')

#Paths for GT and predictions
gt_path = '...'
pred_path = '...'

#Get the file names for evaluation
gt_files = []
for f in os.listdir(gt_path):
    if '.mat' in f:
        gt_files.append(f)
pred_files = []
for f in os.listdir(pred_path):
    if '.mat' in f:
        pred_files.append(f)

#Sort the file names
gt_files.sort(cmp=locale.strcoll)
pred_files.sort(cmp=locale.strcoll)

n_pxls = 0
#Initialize cumulative metric values
crms = 0
crmsl = 0
cabs_rel = 0
csq_rel = 0
cacc1 = 0
cacc2 = 0
cacc3 = 0

for i in xrange(0, len(gt_files)):
    #Load GT and prediction depth maps
    gt = sio.loadmat(os.path.join(gt_path, gt_files[i]))
    pred = sio.loadmat(os.path.join(pred_path, pred_files[i]))
    gt_depths = gt["depth"]
    h, w = np.shape(gt_depths)
    pred_depths = np.squeeze(pred['mat'], axis=[0, 3])
    #Resize prediction to match GT size
    pred_depths = cv2.resize(pred_depths, (w, h), interpolation=cv2.INTER_LINEAR)
    #Cap predicted depth, either [1, 80] or [1, 50]
    pred_depths[pred_depths > 80.0] = 80.0
    pred_depths[pred_depths < 1.0] = 1.0

    #Show the number of evaluated frames
    if i % 100 ==0:
        print(i)

    #Create mask of valid pixels
    mask = gt_depths > 0.01
    crop = np.zeros_like(pred_depths)
    crop[h-219:h-3, 44:1180] = 1
    mask = np.logical_and(mask, (crop == 1))

    #Count number of valid pixels
    n_pxls += np.sum(mask)

    #Compute RMSE
    rms = np.square(gt_depths - pred_depths)
    rms[~mask] = 0
    crms += np.sum(rms)

    gt_depths[gt_depths < 0.5] = 0.5 #epsilon

    #Compute RMSE log
    rmsl = np.square(np.log(gt_depths) - np.log(pred_depths))
    rmsl[~mask] = 0
    crmsl += np.sum(rmsl)

    #Compute ARD
    abs_rel = np.abs(gt_depths - pred_depths) / gt_depths
    abs_rel[~mask] = 0
    cabs_rel += np.sum(abs_rel)

    #Compute SRD
    sq_rel = np.square(gt_depths - pred_depths) / gt_depths
    sq_rel[~mask] = 0
    csq_rel += np.sum(sq_rel)

    max_ratio = np.maximum(gt_depths / pred_depths, pred_depths / gt_depths)

    #Compute accuracies for different deltas
    acc1 = np.asarray(np.logical_and(max_ratio < 1.25, mask), dtype = np.float32)
    acc2 = np.asarray(np.logical_and(max_ratio < 1.25 ** 2, mask), dtype = np.float32)
    acc3 = np.asarray(np.logical_and(max_ratio < 1.25 ** 3, mask), dtype = np.float32)

    cacc1 += np.sum(acc1)
    cacc2 += np.sum(acc2)
    cacc3 += np.sum(acc3)
    

RMSE = np.sqrt(crms / n_pxls)
RMSE_log = np.sqrt(crmsl / n_pxls)
ABS_REL = cabs_rel / n_pxls
SQ_REL = csq_rel / n_pxls
ACCURACY1 = cacc1 / n_pxls
ACCURACY2 = cacc2 / n_pxls
ACCURACY3 = cacc3 / n_pxls

#Display metrics
print(RMSE)
print(RMSE_log)
print(ABS_REL)
print(SQ_REL)
print(ACCURACY1)
print(ACCURACY2)
print(ACCURACY3)
print(n_pxls)
